rs/registry/canister/src/mutations/node_management/do_replace_operator.rs

```diff
@@ -0,0 +1,829 @@
+    use ic_types::{crypto::KeyPurpose, CanisterId, NodeId, PrincipalId, ReplicaVersion};
+
+    use crate::registry::Registry;
+    use prost::Message;
+
+    fn operator(n: u64) -> PrincipalId {
+        PrincipalId::new_user_test_id(n)
+    }
+
+    // To differentiate between `operator(1)` and `provider(1)`.
+    fn provider(n: u64) -> PrincipalId {
+        PrincipalId::new_user_test_id(u64::MAX - n)
+    }
+
+    // Convenience function for readability of
+    // the test.
+    fn caller(n: u64) -> PrincipalId {
+        provider(n)
+    }
+
+    fn node(n: u64) -> NodeId {
+        NodeId::new(PrincipalId::new_node_test_id(n))
+    }
+
+    fn payload(
+        old_operator_id: PrincipalId,
+        new_operator_id: PrincipalId,
+        node_ids: &[NodeId],
+    ) -> ReplaceNodeOperatorPayload {
+        ReplaceNodeOperatorPayload {
+            node_ids: node_ids.to_vec(),
+            new_operator_id,
+            old_operator_id,
+        }
+    }
+
+    trait AssertHelpers {
+        #[track_caller]
+        fn assert_err_contains(self, expected: &str);
+
+        #[track_caller]
+        fn assert_ok(self);
+    }
+
+    impl<T> AssertHelpers for Result<T, String> {
+        #[track_caller]
+        fn assert_err_contains(self, expected: &str) {
+            match self {
+                Ok(_) => panic!("Expected error, but got Ok."),
+                Err(e) => assert!(
+                    e.contains(expected),
+                    "Expected error containing '{expected}', but got '{e}'"
+                ),
+            }
+        }
+
+        #[track_caller]
+        fn assert_ok(self) {
+            assert!(
+                self.is_ok(),
+                "Expected Ok, but got Err: {}",
+                self.err().unwrap()
+            )
+        }
+    }
+
+    fn mutation(key: Vec<u8>, value: Vec<u8>) -> RegistryMutation {
+        RegistryMutation {
+            mutation_type: Type::Upsert as i32,
+            key,
+            value,
+        }
+    }
+
+    fn operator_mutation(
+        operator: PrincipalId,
+        provider: PrincipalId,
+        node_allowance: u64,
+        dc_id: &str,
+    ) -> RegistryMutation {
+        let operator_record = NodeOperatorRecord {
+            node_operator_principal_id: operator.as_slice().to_vec(),
+            node_allowance,
+            node_provider_principal_id: provider.as_slice().to_vec(),
+            dc_id: dc_id.to_string(),
+            ..Default::default()
+        };
+
+        mutation(
+            make_node_operator_record_key(operator).as_bytes().to_vec(),
+            operator_record.encode_to_vec(),
+        )
+    }
+
+    thread_local! {
+        /// Needed for invariant checks of each node
+        /// because each node has to have unique xnet
+        /// and http endpoints
+        static NODE_NUM: RefCell<u8> = const { RefCell::new(0) };
+    }
+
+    fn node_mutation(node_id: NodeId, operator: PrincipalId) -> RegistryMutation {
```

Comment:
This name is under-specified, as it doesn't say what kind of mutation.  I second Daniel's suggestion.

I would also consider moving this fucntion to src/common/test_helpers and adding a couple fields (see the way nodes are made in `prepare_registry_with_nodes_and_node_operator_id` - could easily do that in a function, as it's very similar to what you're doing here).

Reply:
Still left the function here for easier reading of the test, but the side effects have been removed.